### PR TITLE
Gemini 2.0

### PR DESCRIPTION
**Pull request by zeroasterisk**
Update llms.mdx (Gemini 2.0)
- Add Gemini 2.0 flash to Gemini table.
- Add link to 2 hosting paths for Gemini in Tip.
- Change to lower case model slugs vs names, user convenience.
- Add https://artificialanalysis.ai/ as alternate leaderboard.
- Move Gemma to "other" tab.

Update llm.py (gemini 2.0)
- Add setting for Gemini 2.0 context window to llm.py

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comments for crewAI PR #1773

## Overview
This pull request introduces documentation updates and code implementations related to the Gemini 2.0 model, specifically focusing on context window settings and improved model descriptions. While the changes made are largely commendable, there are areas for improvement to enhance clarity, consistency, and maintainability.

## Documentation Changes (`llms.mdx`)

### Positive Aspects
- Comprehensive details about the Gemini 2.0 model have been added.
- Model naming consistency has been improved by standardizing to lowercase.
- New resources, such as **artificialanalysis.ai**, enhance the comparison of models.
- Clear paths for hosting the Gemini model improve user understanding.
- Reorganization of the Gemma model into the "other" section helps in structuring information effectively.

### Suggestions for Improvement
1. **Model Version Consistency**: 
   - Ensure that the model name mentioned in the documentation matches precisely with that in the implementation. For example, `gemini-2.0-flash-exp` is utilized in the code but not reflected in the documentation.

2. **Token Count Formatting**: 
   - Standardize token count presentation across documentation. Consider using exact numeric counts consistently to avoid confusion, i.e., using `2,097,152 tokens` rather than `2M tokens` where necessary.

3. **Addition of Version Information**: 
   - Noting the experimental phase of Gemini 2.0 in the documentation should guide users regarding reliability. For example:
     ```markdown
     <Tip>
     Note: Gemini 2.0 features are currently in the experimental phase as of December 2024. 
     Google's Gemini models are all multimodal...
     </Tip>
     ```

## Code Changes (`llm.py`)

### Positive Aspects
- New model context window settings are added clearly, integrating well with existing code formatting.
- The token value (1048576) is accurately aligned with the documentation.

### Issues and Suggestions
1. **Model Name Consistency**:
   - Align the model names in both the code and the documentation. The entry should be `gemini-2.0-flash` in code for consistency with the documentation.

2. **Missing Constants Documentation**:
   - Missing type hints and docstrings for `CONTEXT_WINDOW_SIZES` hinder clarity. Suggested enhancement:
     ```python
     CONTEXT_WINDOW_SIZES: Dict[str, int] = {
         """
         Dictionary mapping model identifiers to their maximum context window sizes in tokens.
         Updates:
         - Added Gemini 2.0 Flash support with 1M token context window
         """
         # existing entries...
     }
     ```

3. **Consider Adding Version Check**:
   - Implementing validations or warnings for Gemini 2.0, given its experimental nature, could enhance user awareness of potential instability:
     ```python
     if model_name.startswith("gemini-2.0") and not hasattr(self, "_gemini_2_warning_shown"):
         warnings.warn("Gemini 2.0 support is experimental", RuntimeWarning)
         self._gemini_2_warning_shown = True
     ```

## General Recommendations
1. **Testing Requirements**:
   - Comprehensive unit tests are recommended for the new context window settings to ensure full integration and functionality.

2. **Documentation Sync**:
   - Maintain continuous alignment between code comments and documentation updates to guarantee coherence and ease of understanding.

3. **Version Management**:
   - Include notes regarding version compatibility of models and document any breaking changes or deprecations clearly to alert users of potential impacts.

4. **Error Handling**:
   - Implement specific error messages and validation for issues related to experimental features to improve debugging and user experience.

In summary, while the changes in PR #1773 demonstrate significant advancements, addressing the consistency and documentation issues will greatly enhance the user experience and maintainability of the project. Overall, it suggests a strong direction towards improving Gemini 2.0 support within the crewAI framework.

**Review by bhancockio — APPROVED** [review]
Thank you @zeroasterisk for updating the Gemini section!
